text/0000-default-field-values.md

```diff
@@ -0,0 +1,1658 @@
+
+```rust
+pub mod foo {
+    pub struct Alpha {
+        beta: u8 = 42,
+        gamma: bool = true,
+    }
+}
+
+mod bar {
+    fn baz() {
+        let x = Alpha { .. };
+    }
+}
+```
+
+Despite `foo::bar` being in a different module than `foo::Alpha` and despite
+`beta` and `gamma` being private to `foo::bar`, a Rust compiler will accept
+the above snippet. It is legal because when `Alpha { .. }` expands to
+`Alpha { beta: 42, gamma: true }`, the fields `beta` and `gamma` are considered
+in the context of `foo::Alpha`'s *definition site* rather than `bar::baz`'s
+definition site.
+
+## `#[non_exhaustive]` interactions
+
+[RFC 2008]: https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#structs-1
+
+[RFC 2008] introduced the attribute `#[non_exhaustive]` that can be placed
+on `struct`, `enum`, and `enum` variants. The RFC notes that upon defining
+a `struct` in *crate A* such as (12):
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub width: u16,
+    pub height: u16,
+}
+```
+
+it is **_not_** possible to initialize a `Config` in a different *crate B* (13):
+
+```rust
+let config = Config { width: 640, height: 480 };
+```
+
+This is forbidden when `#[non_exhaustive]` is attached because the purpose of
+the attribute is to permit adding fields to `Config` without causing a
+breaking change. However, the RFC goes on to note that you can pattern match
+if you allow for the possibility of having fields be ignored with `..` (14):
+
+```rust
+let Config { width, height, .. } = config;
+```
+
+Because this RFC gives a way to have default field values, you can now simply
+invert the pattern expression and initialize a `Config` like so (15):
+
+```rust
+let config = Config { width, height, .. };
+```
+
+## Defaults for `enum`s
+
+The ability to give fields default values is not limited to `struct`s.
+Fields of `enum` variants can also be given defaults (16):
+
+```rust
+enum Ingredient {
+    Tomato {
+        color: Color = Color::Red,
+        taste: TasteQuality,
+    },
+    Onion {
+        color: Color = Color::Yellow,
+    }
+}
+```
+
+Given these defaults, you can then proceed to initialize `Ingredient`s
+as you did with `struct`s (17):
+
+```rust
+let sallad_parts = vec![
+    Ingredient::Tomato { taste: Yummy, .. },
+    Ingredient::Tomato { taste: Delicious, color: Color::Green, },
+    Ingredient::Onion { .. },
+];
+```
+
+Note that `enum` variants have public fields and in today's Rust,
+this cannot be controlled with visibility modifiers on variants.
+
+Furthermore, when `#[non_exhaustive]` is specified directly on an `enum`,
+it has no interaction with the defaults values and the ability to construct
+variants of said enum. However, as specified by [RFC 2008], `#[non_exhaustive]`
+is permitted on variants. When that occurs, the behaviour is the same as if
+it had been attached to a `struct` with the same fields.
+
+### Specifying the `#[default]` variant
```

Comment:
As I mentioned elsewhere, this seems like an entirely orthogonal feature and could possibly be made into a separate RFC altogether-- there's no particularly interesting interaction between the two AFAICT

Reply:
Yes, I expressed this to Centril in chat and again briefly in my review here. I'm not personally against it being in the same RFC, but it could certainly be factored out into a separate one without any harm, in my view. You guys are on the lang team though, so you decide. :-)

Reply:
Sure; it could be extracted out but I'd like to propose them together first since once you have field defaults in `enum` variants, it's a bit sad if you cannot then `#[derive(Default)]` and make use of those field defaults.